service/resource/configmap/update.go

```diff
@@ -2,12 +2,64 @@ package configmap
 
 import (
 	"context"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/pkg/api/v1"
 )
 
 func (r *Resource) GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
+	currentStateConfigMap, ok := currentState.(*v1.ConfigMap)
+	if !ok {
+		return nil, nil, nil, configMapAssertionError
+	}
+
+	desiredStateConfigMap, ok := desiredState.(*v1.ConfigMap)
+	if !ok {
+		return nil, nil, nil, configMapAssertionError
+	}
+
+	// If the current or desired state configmaps are empty,
+	// perform no action.
+	if currentStateConfigMap == nil {
+		return nil, nil, nil, nil
+	}
+	if desiredStateConfigMap == nil {
+		return nil, nil, nil, nil
+	}
+
+	// If the current and desired state configmaps have different names or namespaces,
+	// something bad is going on, so error out.
+	if currentStateConfigMap.Name != desiredStateConfigMap.Name {
+		return nil, nil, nil, configMapWrongNameError
+	}
+	if currentStateConfigMap.Namespace != desiredStateConfigMap.Namespace {
+		return nil, nil, nil, configMapWrongNamespaceError
+	}
```

Comment:
Personal feedback based on my experiences implementing resources. Such code is easier understandable on a first look when the management of state computation is put into blocks that assign values to variables. See https://github.com/giantswarm/flannel-operator/blob/f6b02e2cc07a364f6d325bdd83e49c4e10a5c806/service/resource/networkconfig/update.go. Here it is easily observable that the create and delete state is empty. In your code I need to scan all conditions down the road to realize the same fact.

Reply:
I just tried this out, I couldn't get it in a way that I liked. Possibly because we only have one case where we actually need to update? Uncertain.

If not a blocker, I would like to proceed with this.

Reply:
This is my point, there is only one case for the update but the code looks like it handles 6 different cases because of all the different conditions. Roll how you like.